### DISCUSSION TITLE
Equilisation and Volume in snapclient

### DISCUSSION THREAD

**Discussion by geoffw64** [pschatzmann/arduino-snapclient#23] [Q&A] [answered]
Hi, 
I'm using this library on an ESP32.  The snapserver is a Pi4.  Music sounds good using opus and a vanilla I2SStream (no wrappers).  
A question.... Is it possible to enable equilisation and volume control with the snapclient? 
Wrapping the I2SStream in a VolumeStream does set the volume, however the doLoop() in snapclient doesn't return regularly, so how would you update the volume while music is playing?  
For equalisation I tried wrapping VolumeStream in a Equilized3Bands stream, however the audio was all broken up.  
Or is the assumption that the volume and equalisation are controlled by a player on the server side?

**Comment by pschatzmann** [comment, marked as answer]
Volume handling and update is part of the snap protocol: so you could e.g. install a mobile app and control the volume via it.
If you want the loop to be called more often, you would need to redesign the SnapProcessor processLoopStep() method or extend the functionality by adding some callback that is called in the loop of  processLoopStep.

**Reply by pschatzmann** [reply]
You could also do the volume handling or the copy in a separate task

**Reply by pschatzmann** [reply]
To minimize the processing time of the loop, I have implemented a DRAFT processLoopStepFast(). This can be activated by calling   client.snapProcessor().setFastLoop(true);

**Reply by geoffw64** [reply]
Thanks Phil, 
I'll try it out.  Appreciate your help.
